pkg/api/cluster_api.go

```diff
@@ -50,10 +50,29 @@ func (o *ClusterDeleteOpts) Validate() error {
 	)
 }
 
+// ClusterSecurityGroupIDGetOpts specifies the required inputs for getting the cluster's security group
+type ClusterSecurityGroupIDGetOpts struct {
+	ID ID
+}
+
+// ClusterSecurityGroupID contains an EKS cluster's cluster security group ID
+// See https://docs.aws.amazon.com/eks/latest/userguide/sec-group-reqs.html
+type ClusterSecurityGroupID struct {
+	Value string
+}
+
+// Validate the delete inputs
+func (o *ClusterSecurityGroupIDGetOpts) Validate() error {
+	return validation.ValidateStruct(o,
+		validation.Field(&o.ID, validation.Required),
+	)
+}
+
 // ClusterService provides an interface for the business logic when working with clusters
 type ClusterService interface {
 	CreateCluster(context.Context, ClusterCreateOpts) (*Cluster, error)
 	DeleteCluster(context.Context, ClusterDeleteOpts) error
+	GetClusterSecurityGroupID(context.Context, ClusterSecurityGroupIDGetOpts) (*ClusterSecurityGroupID, error)
```

Comment:
Thinking back on the discussion about the problem with thingsTM having to implement all kinds of strange different things.

The recommended way of structuring interfaces in Go is to divide functionality into smaller interfaces. In this case adding an interface for cluster crud operations, ClusterCruder, and an interface f.ex ClusterDetailer for functionality like `GetClusterSecurityGroupID`. ClusterService would then extend on both of them. That way functionality only requiring cluster crud operations could accept the ClusterCruder interface instead of the whole service, and the service could act as a ClusterCruder.

That said, I feel the right approach here would actually be to add the security group ID to the cluster struct, and implement GetCluster() functionality.

Reply:
> That said, I feel the right approach here would actually be to add the security group ID to the cluster struct, and implement GetCluster() functionality.

Yeah, I wouldn't mind having that function. I did took a look at how to implement GetCluster, though. That function would return an `api.Cluster` struct, which contains lots of data, which we currently have in state. We're trying not to use state, so a GetCluster should probably retrieve all the info from AWS. It would need that actually, since the state doesn't contain the cluster security group ID, needed by this PR. (Or we could do both, merging some from state and the cluster security group ID from AWS, but that's a bit messy.)

I don't have the full picture, but I think this needs to be done:
* Implement GetCluster to get all info in the `api.Cluster` struct from AWS API calls. (`aws eks describe-cluster` gives a some of the data, but not all.)
* Replace calls to cluster_client.go -> ClusterState.{GetCluster,HasCluster,...), about 15 calls. Perhaps this could be done later, but I don't want us to have multiple sources of truth.
* Consider caching GetCluster results if possible for short amount of time, it might be a bit resource intensive to fetch all that info for every call.

So that's a bit of work. For something as central as GetCluster I see the point of implementing it, though. So if we want to prioritize it, I can give it a go. However, I think it should be done in a separate PR for cleaner separation, and since this solution is working.

> The recommended way of structuring interfaces in Go is to divide functionality into smaller interfaces.

Yeah, will do.

Reply:
> Replace calls to cluster_client.go -> ClusterState.{GetCluster,HasCluster,...), about 15 calls. Perhaps this could be done later, but I don't want us to have multiple sources of truth.

I'd definitely save this for later. Totally agree with not wanting to have multiple ways of doing the same thing, but I'd rather eat that elephant one bite at a time. Each function we implement that directly acquires information from the API saves us strange bugs down the line. IMO

> Consider caching GetCluster results if possible for short amount of time, it might be a bit resource intensive to fetch all that info for every call.

We probably need caching at some point yes, but for now we're ~5~ teams with what, 10 calls to the API each? My suggestion of a cache implementation would be to implement a `api.CachedClusterClient` which took the original cluster client and memoized its functions

> So that's a bit of work. For something as central as GetCluster I see the point of implementing it, though. So if we want to prioritize it, I can give it a go. However, I think it should be done in a separate PR for cleaner separation, and since this solution is working.

I'd say don't prioritize it unless you want to experiment. The thoughts I've communicated here are my default approach to issues like this, and therefore doesn't cost as much time. My suggestion is either to do it this way the next time we touch similar functionality, or if you disagree, discuss it with me. 

> However, I think it should be done in a separate PR for cleaner separation, and since this solution is working.

Totally agree.